src/crimson/os/seastore/transaction_manager.h

```diff
@@ -424,6 +424,140 @@ class TransactionManager : public ExtentCallbackInterface {
     });
   }
 
+  /**
+   * overwrite_pin
+   *
+   * Overwrite extent inside the original extent.
+   * Return the tuple of splited extents and overwrite extent.
+   * One of left pin or right pin can be nullptr.
+   */
+  using overwrite_pin_iertr = base_iertr;
+  template <typename T>
+  using overwrite_pin_ret = overwrite_pin_iertr::future<
+    std::tuple<LBAMappingRef, TCachedExtentRef<T>, LBAMappingRef>>;
+  template <typename T>
+  overwrite_pin_ret<T> overwrite_pin(
+    Transaction &t,
+    LBAMappingRef &&pin,
+    extent_len_t overwrite_offset,
+    extent_len_t overwrite_len,
+    ceph::bufferlist bl) {
+    LOG_PREFIX(TransactionManager::overwrite_pin);
+    // FIXME: paddr can be absolute and pending
+    ceph_assert(pin->get_val().is_absolute());
+
+    auto original_laddr = pin->get_key();
+    auto original_paddr = pin->get_val();
+    auto original_length = pin->get_length();
+    auto l_laddr = original_laddr;
+    auto l_paddr = original_paddr;
+    auto l_len = overwrite_offset;
+    auto o_laddr = original_laddr + overwrite_offset;
+    auto o_paddr = original_paddr.add_offset(overwrite_offset);
+    auto o_len = overwrite_len;
+    auto r_laddr = original_laddr + l_len + o_len;
+    auto r_paddr = original_paddr.add_offset(l_len + o_len);
+    auto r_len = original_length - l_len - o_len;
+    SUBDEBUGT(seastore_tm, "original laddr: {}, paddr: {}, length: {}"
+      " left new laddr: {}, paddr: {}, length: {}"
+      " overwrite new laddr: {}, paddr: {}, length: {}"
+      " right new laddr: {}, paddr: {}, length: {}",
+      t, original_laddr, original_paddr, original_length,
+      l_laddr, l_paddr, l_len,
+      o_laddr, o_paddr, o_len,
+      r_laddr, r_paddr, r_len);
+    ceph_assert(original_length > o_len);
+    ceph_assert(o_laddr >= original_laddr);
+    ceph_assert(o_len == bl.length());
+    ceph_assert(r_laddr <= original_laddr + original_length);
+
+    return seastar::do_with(
+      std::move(bl),
+      CachedExtentRef(),
+      LBAMappingRef(),
+      LBAMappingRef(),
+      [=, this, &t](auto &bl, auto &oext, auto &lpin, auto &rpin) mutable {
+      return cache->get_extent_if_cached(
+        t, original_paddr, T::TYPE
+      ).si_then([=, this, &t, &oext](auto ext) {
+        ceph_assert(ext ? ext->is_clean() : true);
```

Comment:
Why ext->is_clean() should be true?

Reply:
it should be `ceph_assert(ext->state == CachedExtent::extent_state_t::CLEAN);`  
this is to make sure not to remap a dirty extent, which should be MERGE_EXISTING cases in overwrite

Reply:
@aisakaki  is_pending() is not related to whether the extent is dirty.

Reply:
Same applies to another place checking is_clean()